### PR TITLE
funfile: handle disabled download links. resolves #7436

### LINKED ISSUE #7436: [funfile] Exception (funfile): The string did not match the expected pattern.: Parse error (Test)
**Please use the search bar** at the top of the page and make sure you are not creating an already submitted issue.
Check closed issues as well, because your issue may have already been fixed.

Please read our [Contributing Guidelines](https://github.com/Jackett/Jackett/blob/master/CONTRIBUTING.md) before submitting your issue to ensure a prompt response to your bug.

### Environment

**OS**: CentOS 7

**.Net Runtime**: [.Net-Core/.Net-Framework/Mono]

**.Net Version**:

**Jackett Version**:  0.13.378.0

**Last Working Jackett Version**: Not sure, was working yesterday.

**Are you using a proxy or VPN?** no
### Description

Test FunFile connection and fails

### Logged Error Messages

Jackett.Common.IndexerException: Exception (funfile): The string did not match the expected pattern.
 ---> System.Exception: Parse error
 ---> AngleSharp.Dom.DomException: The string did not match the expected pattern.
   at AngleSharp.Dom.QueryExtensions.CreateSelector(INodeList nodes, INode scope, String selectorText)
   at AngleSharp.Dom.QueryExtensions.QuerySelector(INodeList nodes, String selectorText, INode scopeNode)
   at AngleSharp.Dom.Element.QuerySelector(String selectors)
   at Jackett.Common.Indexers.FunFile.PerformQuery(TorznabQuery query) in /home/vsts/work/1/s/src/Jackett.Common/Indexers/FunFile.cs:line 122

### PR DESCRIPTION
**Pull request by ngosang**
The user will see this message instead of the null pointer => `Download links not found. Make sure you can download from the website.`

### CODE REVIEW

File: src/Jackett.Common/Indexers/FunFile.cs [hunk 1]
```diff
@@ -77,27 +77,17 @@ protected override async Task<IEnumerable<ReleaseInfo>> PerformQuery(TorznabQuer
...
+            var qc = new NameValueCollection
             {
-                cat = cats[0];
-            }
-            queryCollection.Add("cat", cat);
-
-            searchUrl += "?" + queryCollection.GetQueryString();
+                {"incldead", "1"},
+                {"showspam", "1"},
+                {"cat", cats.Count > 0 ? cats[0] : "0"}
```
Comment (cadatoiva) [inline src/Jackett.Common/Indexers/FunFile.cs]:
```suggestion
                {"cat", MapTorznabCapsToTrackers(query).FirstOrDefault() ?? "0"}
```

File: src/Jackett.Common/Indexers/FunFile.cs [hunk 2]
```diff
@@ -77,27 +77,17 @@ protected override async Task<IEnumerable<ReleaseInfo>> PerformQuery(TorznabQuer
...
-            var queryCollection = new NameValueCollection();
-            queryCollection.Add("incldead", "1");
-            queryCollection.Add("showspam", "1");
-
-            if (!string.IsNullOrWhiteSpace(searchString))
-            {
-                queryCollection.Add("search", searchString);
-            }
-
             var cats = MapTorznabCapsToTrackers(query);
```
Comment (cadatoiva) [inline src/Jackett.Common/Indexers/FunFile.cs]: If we use `.FirstOrDefault()` we can avoid creating a variable here.
```suggestion

```

File: src/Jackett.Common/Indexers/FunFile.cs [hunk 3]
```diff
@@ -126,14 +116,17 @@ protected override async Task<IEnumerable<ReleaseInfo>> PerformQuery(TorznabQuer
                     var qTimeAgo = row.QuerySelector("td:nth-of-type(6)");
                     var qSize = row.QuerySelector("td:nth-of-type(8)");
 
-                    var catStr = qCatLink.GetAttribute("href").Split('=')[1].Split('&')[0];
-                    release.Category = MapTrackerCatToNewznab(catStr);
+                    if (qDownloadLink == null)
+                        throw new Exception("Download link not found. Make sure you can download from the website.");
```
Comment (cadatoiva) [inline src/Jackett.Common/Indexers/FunFile.cs]: If the only time a download link is expecting to be missing is when they're disabled for every torrent in the tracker, then this is ok, but the exception should reflect this as below. If only some of the results don't have a download link for whatever reason (like 720pier is a forum with most, not all, posts including torrent downloads as example), then we should be using `continue` instead.
```suggestion
                    if (qDownloadLink == null)
                        throw new Exception("Download links not found. Make sure you can download from the website.");
```

File: src/Jackett.Common/Indexers/FunFile.cs [hunk 4]
```diff
@@ -126,14 +116,17 @@ protected override async Task<IEnumerable<ReleaseInfo>> PerformQuery(TorznabQuer
...
 
                     release.Link = new Uri(SiteLink + qDownloadLink.GetAttribute("href"));
                     release.Title = qDetailsLink.GetAttribute("title").Trim();
                     release.Comments = new Uri(SiteLink + qDetailsLink.GetAttribute("href"));
                     release.Guid = release.Link;
 
+                    var catStr = qCatLink.GetAttribute("href").Split('=')[1].Split('&')[0];
+                    release.Category = MapTrackerCatToNewznab(catStr);
```
Comment (cadatoiva) [inline src/Jackett.Common/Indexers/FunFile.cs]: I know I just mentioned it in an already closed PR, but some variation on the code below is less error prone if the url format changes.
```
var qParams = HttpUtility.ParseQueryString(new Uri(qCatLink.GetAttribute("href")).Query);
var catStr = qParams["catParam"];
```

File: src/Jackett.Common/Indexers/FunFile.cs (hunk 2 above)
Comment (ngosang) [reply src/Jackett.Common/Indexers/FunFile.cs]: done

File: src/Jackett.Common/Indexers/FunFile.cs (hunk 1 above)
Comment (ngosang) [reply src/Jackett.Common/Indexers/FunFile.cs]: done

File: src/Jackett.Common/Indexers/FunFile.cs (hunk 3 above)
Comment (ngosang) [reply src/Jackett.Common/Indexers/FunFile.cs]: In this case the user will see all download links or none.

File: src/Jackett.Common/Indexers/FunFile.cs (hunk 4 above)
Comment (ngosang) [reply src/Jackett.Common/Indexers/FunFile.cs]: If you take a look at the previous lines, we are already hardcoding that cat is the first query param.
`var qCatLink = row.QuerySelector("a[href^=\"browse.php?cat=\"]");`
I didn't change that code, it's not the goal of this PR.

**Pull request merged**